regolith/helpers/l_contactshelper.py

```diff
@@ -0,0 +1,167 @@
+        rc = self.rc
+        if "groups" in self.needed_dbs:
+            rc.pi_id = get_pi_id(rc)
+        rc.coll = f"{TARGET_COLL}"
+        try:
+            if not rc.database:
+                rc.database = rc.databases[0]["name"]
+        except BaseException:
+            pass
+        colls = [
+            sorted(
+                all_docs_from_collection(rc.client, collname), key=_id_key
+            )
+            for collname in self.needed_dbs
+        ]
+        for db, coll in zip(self.needed_dbs, colls):
+            gtx[db] = coll
+        gtx["all_docs_from_collection"] = all_docs_from_collection
+        gtx["float"] = float
+        gtx["str"] = str
+        gtx["zip"] = zip
+
+    def sout(self):
+        rc = self.rc
+        contacts = []
+        ret_list = []
+        num = 0
+        if rc.name:
+            num += 1
+            for name in rc.name:
+                contacts.extend(
+                    fragment_retrieval(
+                        self.gtx['contacts'], [
+                            "_id", "aka", "name"], name))
+        if rc.inst:
+            num += 1
+            contacts.extend(
+                fragment_retrieval(
+                    self.gtx['contacts'],
+                    ["institution"],
+                    rc.inst))
+        if rc.notes:
+            num += 1
+            for note in rc.notes:
+                contacts.extend(
+                    fragment_retrieval(
+                        self.gtx['contacts'],
+                        ["notes"],
+                        note))
+        if rc.date:
+            num += 1
+            temp_dat = date_parser.parse(rc.date).date()
+            for contact in self.gtx["contacts"]:
+                if rc.date:
+                    if rc.range:
+                        temp_dict = {
+                            "begin_date": temp_dat -
+                            dateutil.relativedelta.relativedelta(
+                                months=int(
+                                    rc.range)),
+                            "end_date": temp_dat +
+                            dateutil.relativedelta.relativedelta(
+                                months=int(
+                                    rc.range))}
+                    else:
+                        temp_dict = {
+                            "begin_date": temp_dat -
+                            dateutil.relativedelta.relativedelta(
+                                months=2),
+                            "end_date": temp_dat +
+                            dateutil.relativedelta.relativedelta(
+                                months=2)}
+                    if is_current(temp_dict, now=temp_dat):
+                        contacts.append(contact)
+        for con in contacts:
+            ret_list.append(
+                f"name: {con.get('name')}, institution: {con.get('institution')}, email: {con.get('email','missing')}")
+        temp_c = Counter(ret_list)
+        for key in temp_c:
```

Comment:
What is the intention of this block?

Reply:
This is what ensures there's AND logic in the code. I'll explain it using an example. Let's say I enter "regolith helper l_contacts -n sh -i col -o bill -d 2020-06-06"

First, I go through each field (name, institution, note, date). For each field, I search the database and add every contact in the database that has the corresponding fragment in it (using the fragment_retrieval function). I can't just call fragment_retrieval on all four fields at the same time because I only want institutions with "col" in them (I don't want institutions with "sh" or "bill" in them).

In this example, I entered all four fields, so I my "contacts" list should now have at least four entries, all of which are duplicates and all of which correspond to the entry I'm looking for in the database. There might also be other documents in the contacts list, but those will not occur four times in the list (so I can filter them out in my next step). These are documents that only corresponded to some, but not all of the entries (so for example a contact that has a name containing "sh" but did not attend an institution containing "col").

Then, I use the "Counter" class to calculate how many times each entry in the "contacts" list appears and store that in a dictionary called "temp_c".

Then, I know to only print the entries in temp_c that appear as many times as the number of fields I initially entered (represented by the "num" variable).

This ensures AND logic by making sure that every contact I return corresponded to every single field I entered.